frontend-wasm/src/code_translator/mod.rs

```diff
@@ -0,0 +1,1035 @@
+use miden_hir::Type::*;
+use miden_hir::{Block, Inst, InstBuilder, Value};
+use wasmparser::{MemArg, Operator};
+
+#[cfg(test)]
+mod tests;
+
+#[cfg(test)]
+mod tests_unsupported;
+
+/// Translates wasm operators into Miden IR instructions.
+pub fn translate_operator(
+    op: &Operator,
+    builder: &mut FunctionBuilderExt,
+    state: &mut FuncTranslationState,
+    mod_info: &ModuleInfo,
+    diagnostics: &DiagnosticsHandler,
+    span: SourceSpan,
+) -> WasmResult<()> {
+    if !state.reachable {
+        translate_unreachable_operator(&op, builder, state, mod_info, span)?;
+        return Ok(());
+    }
+
+    // Given that we believe the current block is reachable, the FunctionBuilderExt ought to agree.
+    debug_assert!(!builder.is_unreachable());
+
+    match op {
+        /********************************** Locals ****************************************
+         *  `get_local` and `set_local` are treated as non-SSA variables and will completely
+         *  disappear in the Miden IR
+         ***********************************************************************************/
+        Operator::LocalGet { local_index } => {
+            let val = builder.use_var(Variable::from_u32(*local_index));
+            state.push1(val);
+        }
+        Operator::LocalSet { local_index } => {
+            let val = state.pop1();
+            builder.def_var(Variable::from_u32(*local_index), val);
+        }
+        Operator::LocalTee { local_index } => {
+            let val = state.peek1();
+            builder.def_var(Variable::from_u32(*local_index), val);
+        }
+        /********************************** Globals ****************************************/
+        Operator::GlobalGet { global_index } => {
+            let global_index = GlobalIndex::from_u32(*global_index);
+            let name = mod_info.global_name(global_index);
+            let ty = mod_info.globals[global_index].ty.clone();
+            state.push1(builder.ins().load_symbol(name, ty, span));
+        }
+        Operator::GlobalSet { global_index } => {
+            let global_index = GlobalIndex::from_u32(*global_index);
+            let name = mod_info.global_name(global_index);
+            let ty = (&mod_info.globals[global_index]).ty.clone();
+            let ptr = builder
+                .ins()
+                .symbol_addr(name, Ptr(ty.clone().into()), span);
+            let val = state.pop1();
+            builder.ins().store(ptr, val, span);
+        }
+        /********************************* Stack misc **************************************/
+        Operator::Drop => _ = state.pop1(),
+        Operator::Select => {
+            let (arg1, arg2, cond) = state.pop3();
+            // if cond is not 0, return arg1, else return arg2
+            // https://www.w3.org/TR/wasm-core-1/#-hrefsyntax-instr-parametricmathsfselect%E2%91%A0
+            let zero = builder.ins().i32(0, span);
+            let cond_i1 = builder.ins().neq(cond, zero, span);
```

Comment:
*Suggestion*: As much as possible, prefer the `_imm`-suffixed instruction variants for things like this, it makes the IR more canonical by construction, and avoids having to deal with the immediate value during analysis/transformation. I think it tends to be more efficient in the VM as well, but that's less important.

```suggestion
            let cond_i1 = builder.ins().neq_imm(cond, Immediate::U32(0), span);
```

Reply:
Sure. Thanks for cathing this! Done.